### DISCUSSION TITLE
Custom Telemetry

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#112] [BC Best Practices]
Custom Telemetry Pattern

**Comment by waldo1001** [comment]
Guidelines coming from @KennieNP :
- Only Telemetry that is "actionable" by the partner (not "nice to know")
- Consider cost-aspects
- Must be documented
- Do not break 

Coding patterns, including 
- custom dimensions
- Naming  conventions 
- ...

May be @KennieNP can take the lead, as this is already documented somewhere internally at Microsoft.

**Comment by KennieNP** [comment]
**This is a copy-paste from our internal guidelines**

=== Guiding principles for telemetry ===

==== Think about it as an API ====
Customers will build analytics and monitoring solutions on top of their telemetry data. 

Therefore, signal must be treated as any other API
* documented (preferably in docs)
* versioned
* discoverable (preferably in docs)
* non-breaking

==== Naming conventions and telemetry schema ====
To make it easy for the consumer of telemetry to work with the data, please 
* use PamelCasing. This makes all fields in Application Insights look the same (signal logged through the AL LogMessage method will have "al" prefixed to dimension names.
* never use fields/custom dimension keys with spaces or special characters. This makes the KQL queries so much easier to write
* for custom dimensions, consider using prefixes that helps the telemetry consumer understand where the dimension is coming from (e.g. HttpStatusCode, SqlStatement, ...)

Consider always having a "message" field that expresses in human readable form what the telemetry event is about. 
If you do, let message names follow the Object ActionInPastTense pattern
Some examples
* Web Service Called: <endpoint> 
* Email attempt failed 
* Authorization to environment <environment name> succeeded


==== Candidate data for telemetry ====
Telemetry must be actionable for the customer. Do not emit signal that they cannot act on (knowing about CPU performance counters on the database is useless if the partner cannot scale the database)

Also, note that customers pay for data ingestion. So be mindful to not flood their telemetry resources.

If you do not know where to start, consider using telemetry for deflection. In Dynamics 365 Business Central, they started with signal about authorization (successful/failed) to deflect support cases and IcM tickets that was due to disabled users/wrong licenses, ...

==== How customers use telemetry ====
The following are known scenarios for customer telemetry
* The tenant admin (typically an itpro) wants to troubleshoot a performance problem and they need more details than what is provided in the analytics reports in PPAC
* The customer wants to analyze (and correct) errors happening in the environment (typically an itpro)
* The customer wants to analyze usage of features (typically an analytics user, maybe with BI experience)

Customers typically start in the Application Insights portal and then move on to use more advanced tools for analytics (KQL, Power BI, Excel, ...). Once they have seen the light, they will likely start alerting on telemetry using Azure Monitor Alerts or setting up Power Automate flows. 

Business Central have developed a telemetry maturity model (based on the Gartner BI maturity model) for how organizations can evolve to use telemetry proactively in their business processes.

=== Privacy considerations ===

==== What we NEVER can emit ====
Telemetry must be privacy compliant.

Decorate emitter methods in code with privacy classification (use the MS taxonomy) to help developers think of privacy at development time.

Only emit signal that is of type System Metadata or OII.
